Feature/m0081 meja 1638 document portable fo (#375)
MicroEJ/docs: KernelDeveloperGuide/kernelLink.rst

   Features linked using the Firmware Linker tool cannot be dynamically uninstalled using `Kernel.uninstall(Feature) <https://repository.microej.com/javadoc/microej_5.x/apis/ej/kf/Kernel.html#uninstall-ej.kf.Feature->`_ method.

.. _controlled_portability:

Controlled Portability
----------------------

A Kernel can install ``.fo`` files that have been built on other Kernels, provided this Kernel complies with other Kernels according to a set of rules declared hereafter.

> Such phase is called `Controlled Portability` because the portability is verified during the new Kernel build, with no impact on the Feature dynamic installation.

Author: reworded to `Feature Portability Control`

Principle
~~~~~~~~~

> During a Kernel build, SOAR can verify this Kernel preserves the portability of ``.fo`` files built on a previous Kernel using :ref:`the Kernel metadata file <kernel_metadata_generation>`.

Reviewer: ```suggestion
During a Kernel build, SOAR can verify this Kernel preserves the portability of any ``.fo`` files built on a previous Kernel using :ref:`the Kernel metadata file <kernel_metadata_generation>`.
```

> If the portability is preserved, the Kernel UIDs are embedded in the new Kernel, allowing ``.fo`` files built on previous Kernels to be installed.

Author: doc updated with Kernel UID dedicated section

> If the portability is preserved, the :ref:`Kernel UID <kernel_uid>` is embedded in the new Kernel, allowing ``.fo`` files built on the previous Kernel to be installed as well.

Reviewer 2: the Kernel UID => the UID of the old Kernel

Otherwise, SOAR fails with an error indicating the broken rules.

> .. figure:: png/controlled_portability_principle.png

Reviewer 2: what is the red box in this figure? i thought it was the kernel, but it seems that it includes a green box which is called "Kernel". maybe green box is Kernel Application code?

Reviewer 2: change color of kernel APIs / metadata

Kernel => Kernel Application

Author: updated feature_portability_control schematic

- Kernel API -> moved to GREEN as Kernel Java Code
- Kernel -> renamed to Kernel Application
- Kernel Metadata -> moved to BLUE as debug sections

Reviewer 2: ok

   :alt: Feature Portability Control Principle
   :align: center
   :scale: 80%

   Feature Portability Control Principle

Enable
~~~~~~

.. note::

   This is a new functionality that requires a custom Architecture configuration.
   Please contact :ref:`our support team <get_support>` for more details.

Add ``com.microej.soar.kernel.compatiblemetadata.path`` option to your Kernel :ref:`Application Options <application_options>`.
This is an OS path-separated list of Kernel Metadata files and directories containing ``*.kdat`` files.

Portability Rules
~~~~~~~~~~~~~~~~~

A Kernel Application can install a ``.fo`` file that has been built against an other Kernel Application

> if the Application code has not changed or complies with following rules:

Reviewer: ```suggestion
if the Kernel Application code has not changed or if the modifications respect the portability rules. 
Here is the list of the modifications that can be done while preserving the portability:
```

- Modify method code, except if :ref:`soar_method_devirtualization` or :ref:`soar_method_inlining` has changed.
- Add a new type (including declared as Kernel API),
- Add a new static method (including declared as Kernel API),
- Add a new instance method in a type **not declared** as Kernel API,
- Add a new instance method with ``private`` visibility in a type **declared** as Kernel API,
- Add a new static field (including declared as Kernel API),
- Add a new instance field in a type **not declared** as Kernel API,
- Rename an instance field with ``private`` visibility in a type **declared** as Kernel API,

> - Modify a Java type, method or static field **not declared** as Kernel API (code, signature, hierarchy) 
> - Remove a Java type, method or static field **not declared** as Kernel API

Reviewer: ```suggestion
- Modify a Java type, method, or static field **not declared** as Kernel API (code, signature, hierarchy) 
- Remove a Java type, method, or static field **not declared** as Kernel API
```

Both Kernel Applications must be built from Platforms based on the same Architecture version.

> Any other modifications will break the Feature portability, for example:

Reviewer: ```suggestion
Any other modifications will break the Feature portability. For example, the following modifications will not preserve the portability:
```
